### ISSUE TITLE
pyright not understanding NewType helper function for types

### ISSUE THREAD

**Issue by jwthomp** [microsoft/pyright#368]
**Describe the bug**
pyright is not honoring the NewType() helper function for defaultdict. A direct cast is required.

**To Reproduce**
Run the following example code:

```
from collections import defaultdict
from typing import NewType, DefaultDict, cast

Foop = NewType("Foop", DefaultDict[int, int])

# No type error
dd: DefaultDict[int, int] = defaultdict(int)

# pyright type error
# No overloads for 'Foop(defaultdict(int))' match parameters
# Argument types: (defaultdict[Any, int])
f1: Foop = Foop(defaultdict(int))

# pyright type error
# No overloads for 'Foop(dd)' match parameters
# Argument types: (DefaultDict[int, int])
f2: Foop = Foop(dd)

#
# Now using casts

# No type error
w1: Foop = cast(Foop, defaultdict(int))

# No type error
w3: Foop = cast(Foop, dd)
```

**Expected behavior**
pyright should handle the NewType() generated helper functions and not require the cast.

**VS Code extension or command-line**
```
$ pyright --version
pyright 1.1.1
```

**Additional context**
From the NewType documentation

 https://docs.python.org/3/library/typing.html#newtype
 ```
   from typing import NewType
   UserId = NewType('UserId', int)
   some_id = UserId(524313)
 ```

 and https://docs.python.org/3/library/typing.html#typing.NewType

 ```
   typing.NewType(typ)
   A helper function to indicate a distinct types to a typechecker, see NewType. At runtime it returns a function that returns its argument.
```

Thank you for all of your hard work. pyright is a really nice tool!

**Comment by erictraut** [comment]
I think pyright is doing the right thing in this case. It's creating a new type that is equivalent to the specified type (`DefaultDict[int, int]`). When you instantiate a new DefaultDict, you need to pass it a factory Callable. Typically, this comes from the constructor of a class that you pass.

For example:
```
f1: DefaultDict[int, int] = DefaultDict(int) # Constructor for int provides factory callable
```
The `NewType` call creates a type alias for the specified type, so in your example, `Foop` is equivalent to `DefaultDict[int, int]`. To instantiate it, you need to provide a factory callable that returns an `int` instance. You're passing it an instance of `defaultdict(int)` which isn't callable.

For more details about DefaultDict, check out this documentation: https://docs.python.org/3/library/collections.html#collections.defaultdict

**Comment by jwthomp** [comment]
Hi @erictraut ,

Thanks for your response. I'll take a look to make sure I understand your answer.

I have no doubt that you know far more about the inner workings of Pythons type system than I do, but one point of confusion for me is that my understanding of NewType was that it was not creating a type alias.

For example, if I were creating a type alias `Foo` from the `int` type, then `a: Foo = 1` would pass the type checker as `Foo` and `int` are interchangeable. 

However, NewType looks to be designed to not act as a type alias that is structural, but rather as a nominal type so that `Foo` and `int` are not directly interchangeable. Similar to what you might do with a type constructor `type foo = Foo int` though underneath I'm sure its different.

It looks like this is accomplished by declaring the NewType as a subclass of the base type. From the docs:

```
Note Recall that the use of a type alias declares two types to be equivalent to one another. Doing Alias = Original will make the static type checker treat Alias as being exactly equivalent to Original in all cases. This is useful when you want to simplify complex type signatures.
In contrast, NewType declares one type to be a subtype of another. Doing Derived = NewType('Derived', Original) will make the static type checker treat Derived as a subclass of Original, which means a value of type Original cannot be used in places where a value of type Derived is expected. This is useful when you want to prevent logic errors with minimal runtime cost.
```

I'll definitely read up on DefaultDict so I can full understand your response. Thank you again!

Jeff

**Comment by erictraut** [comment]
You are correct in your interpretation of "NewType". The type it creates is not meant to be be completely interchangeable with the type it is aliasing. It is meant to act as a subclass of that type, so it inherits all of the properties, behaviors, methods, etc. 

You're creating a new type based on DefaultDict, so it takes on all of the properties and behaviors of DefaultDict. This includes its constructor and initialization method. 

The expression `DefaultDict[int, int](defaultdict(int))` is in error from a type perspective, so therefore `Foop(defaultdict(int))` is also considered in error.

**Comment by jwthomp** [comment]
Ok, thank you. I'll dig into the DefaultDict stuff and make sure I understand it then. Thank you again Eric.

**Comment by karolyi** [comment]
@jwthomp, not sure if you're looking for this but after about a day of fiddling around with it, I found out how to make a typed defaultdict that will typecheck and also be syntactically valid (as in, python will not error out upon running it). My use case was that I needed a type that is type checked while I'm coding, and also the type(variable) will represent the type I created.

I did it by simply subclassing a type definition. Here's an example:
```python
class PercentToVatItem(DefaultDict[int, SumVatItem]):
    'Depicting `vat_percent` -> `SumVatItem` assignments.'

    @staticmethod
    def get_empty():
        return PercentToVatItem(SumVatItem.get_empty)

    def dump(self):
        'Return a JSON dumpable format.'
        result: List[_DumpedVatItem] = []
        for vat_percent, vat_item in self.items():
            result.append([vat_percent, vat_item.dump()])
        return result

    @staticmethod
    def from_dump(data: List[_DumpedVatItem]):
        'Return with loaded data from a passed dump.'
        result = PercentToVatItem.get_empty()
        for vat_percent, dump_item in data:
            assert isinstance(vat_percent, int)
            result[vat_percent] = SumVatItem.from_dump(
                data=cast(typ=SumVatItemDump, val=dump_item))
        return result
```

This only applies for `defaultdict` in my code so far, and has the advantage that you can add your customized business logic methods to the subclassed typed `defaultdict`, as you can see it in the example.

I haven't figured out a way to be able to the same with a `TypedDict`, as it can't contain methods (only annotations), so there I went with dataclasses which did the job.
